WalletWasabi.Tests/UnitTests/Crypto/ZeroKnowledge/LinearRelationTests.cs

```diff
@@ -75,35 +76,119 @@ public void IgnoredWitnessComponents()
+			Assert.True(equation.Verify(publicNonce, challenge, response2));
 
 			// With different witnesses the responses should be different even if the
 			// nonces are the same, but since the first part of the witness is the
 			// same that sub-response should be the same for the same nonce
 			Assert.False(response1 == response2);
 			Assert.True(response1.First() == response2.First());
 		}
+
+		[Fact]
+		public void StatementAndKnowledge()
+		{
+			var x = new Scalar(42);
+			var a = x * Generators.Gg;
+			var b = x * Generators.Gh;
+
+			// Discrete log equality (Chaum-Pedersen proof)
+			var statement = new Statement(new[]
+			{
+				new Equation(a, new GroupElementVector(Generators.Gg)),
+				new Equation(b, new GroupElementVector(Generators.Gh)),
+			});
+
+			// Check that simulation produces valid transcripts
+			var challenge = new Scalar(13);
+			var simulatedResponses = new[]
+			{
+				new ScalarVector(Scalar.One),
+				new ScalarVector(Scalar.One),
+			};
+
+			var simulatedPublicNonces = statement.SimulatePublicNonces(challenge, simulatedResponses);
+			Assert.True(statement.CheckVerificationEquation(simulatedPublicNonces, challenge, simulatedResponses));
+
+			// Create transcripts using a witness to the relation
+			var knowledge = new Knowledge(statement, new ScalarVector(x));
+			var allSecretNonces = new[] { new ScalarVector(new Scalar(7)), new ScalarVector(new Scalar(11)) };
+			var publicNonces = new GroupElementVector(statement.Equations.Zip(allSecretNonces, (equation, secretNonces) => secretNonces * equation.Generators));
+			var allResponses = knowledge.RespondToChallenge(challenge, allSecretNonces);
+			Assert.True(statement.CheckVerificationEquation(publicNonces, challenge, allResponses));
+
+			// Ensure that verifier rejects invalid transcripts
+			Assert.False(statement.CheckVerificationEquation(publicNonces, new Scalar(17), allResponses));
+			Assert.False(statement.CheckVerificationEquation(simulatedPublicNonces, challenge, allResponses));
+			Assert.False(statement.CheckVerificationEquation(publicNonces, challenge, simulatedResponses));
+		}
+
+		[Fact]
+		public void StatementThrows()
+		{
+			// Jagged equation matrices are not allowed
+			Assert.ThrowsAny<ArgumentException>(() => new Statement(new[]
+			{
+				new Equation(GroupElement.Infinity, new GroupElementVector(Generators.Gg, Generators.Gh)),
+				new Equation(GroupElement.Infinity, new GroupElementVector(Generators.G)),
+			}));
+
+			// Null generators are not allowed
+			Assert.ThrowsAny<ArgumentException>(() => new Statement(new[]
+			{
+				new Equation(GroupElement.Infinity, new GroupElementVector(Generators.Gg, Generators.Gh)),
+				new Equation(GroupElement.Infinity, new GroupElementVector(Generators.G, null!)),
+			}));
+		}
+
+		[Fact]
+		public void KnowledgeThrows()
+		{
+			var x = new Scalar(42);
+			var a = x * Generators.Gg;
+			var b = x * Generators.Gh;
+
+			var statement = new Statement(new[]
```

Comment:
is it worth it to add a params overload for `Statement` to clean this up a bit?

Reply:
Done